### PR TITLE
Issue/issue 299

### PR DESCRIPTION
**Pull request by jvalls-axa**
# Issue 299 Part I (PdfMiner)

This PR is part I of #299 -> Heap out of memory in Pdminer extraction & dumppdf extraction.

## Changes:
- Use of xml-stream to read xml pdfminer output file
- Pdfs will be extracted using slots of 500 pages
- Allow node objects to be up to 4Gb in memory using `--max-old-space-size=4096` (Temporally fix)

## Improvements:
- Better feedback running huge documents (Logs will be displayed every 500 pages)
- Faster extraction time using PdfMnier + Xml-Stream

## TODO:
- Fix  'dumppdf.py' Heap out of memory

## xml2js vs xml-stream
17 min VS 9 min

XmlStream

### CODE REVIEW

File: server/src/input/pdfminer/PdfminerExtractor.ts [hunk 1]
```diff
@@ -27,17 +30,70 @@ import * as pdfminer from './pdfminer';
...
-      );
+      return this.pageNumber(repairedPdf).then(totalPages => {
+        logger.info(
+          'Extracting contents (' +
+            totalPages.toString() +
+            " pages) with pdfminer's pdf2txt.py tool...",
+        );
+        const startTime: number = Date.now();
+        const extractFont = extractImagesAndFonts(repairedPdf);
+        const pdfminerExtract = this.extractFile(repairedPdf, 1, 500, totalPages);
```
Comment (peter-vandenabeele-axa) [inline server/src/input/pdfminer/PdfminerExtractor.ts]: Maybe place the value of `500` as a clearly visible `constant` at top of file or even a config parameter with a sensible default value ?

**Comment by peter-vandenabeele-axa** [comment]
Nice :-)
* The description of this PR has a typo `PdfMnier`.
* How was the value of `500` determined/optimized (the batch size for PDF parsing)?
Once you have done this effort, to split it in smaller batches, I would think that taking e.g. 100 pages per batch (instead of 500) would further reduce the risk for out-of-memory ?
Maybe making it a config parameter can resolve my question (then a user can further reduce it, if 500 or 100 is still too large for certain Use Cases?)

File: server/src/types/PdfminerPage.ts [hunk 2]
```diff
@@ -25,18 +25,26 @@ export class PdfminerPage {
...
+      this.textbox = jsonObj.textbox.map(tb => new PdfminerTextbox(tb));
+    } else if (jsonObj.textbox != null) {
+      this.textbox = [new PdfminerTextbox(jsonObj.textbox)];
+    }
+    if (Array.isArray(jsonObj.figure)) {
+      this.figure = jsonObj.figure.map(fig => new PdfminerFigure(fig));
+    } else if (jsonObj.figure != null) {
+      this.figure = [new PdfminerFigure(jsonObj.figure)];
+    }
+    /*this.line = jsonObj.line;
```
Comment (peter-vandenabeele-axa) [inline server/src/types/PdfminerPage.ts]: Useful to keep this commented code in the commit ?

File: server/src/types/PdfminerPage.ts [hunk 3]
```diff
@@ -25,18 +25,26 @@ export class PdfminerPage {
   };
   public textbox: PdfminerTextbox[];
   public figure: PdfminerFigure[];
-  public line: object[];
+  /*public line: object[];
```
Comment (peter-vandenabeele-axa) [inline server/src/types/PdfminerPage.ts]: Useful to keep this commented code in the commit ?

**Comment by jvalls-axa** [comment]
> Nice :-)
> 
> * The description of this PR has a typo `PdfMnier`.
> * How was the value of `500` determined/optimized (the batch size for PDF parsing)?
>   Once you have done this effort, to split it in smaller batches, I would think that taking e.g. 100 pages per batch (instead of 500) would further reduce the risk for out-of-memory ?
>   Maybe making it a config parameter can resolve my question (then a user can further reduce it, if 500 or 100 is still too large for certain Use Cases?)

Hi @peter-vandenabeele-axa and thanks for your feedback !! 👍 

I used 500 pages because after a lot of tests with some pdfs is the bigger number of pages supported with no 'out of memory' and no '--max-old-space-size' committed [here](https://github.com/axa-group/Parsr/pull/315/commits/df3e4198494681c51430ada7c8eb24c05659ede6)

Our goal is allow Parsr to run properly with huge Pdfs (2K pages or more) and avoid '--max-old-space-size' usage.

But anyway adding 'batch max pages'  in configuration file has a lot of sense!

**Comment by peter-vandenabeele-axa** [comment]
> I used 500 pages because after a lot of tests with some pdfs is the bigger number of pages supported with no 'out of memory' and no '--max-old-space-size' committed [here](https://github.com/axa-group/Parsr/pull/315/commits/df3e4198494681c51430ada7c8eb24c05659ede6)
> 
> Our goal is allow Parsr to run properly with huge Pdfs (2K pages or more) and avoid '--max-old-space-size' usage.
> 
> But anyway adding 'batch max pages' in configuration file has a lot of sense!

Yes, I would suggest so. I am a bit afraid that other users or use cases will pop up where the "complexity per page" will be larger, so it will get OOM at smaller amount of pages.

I vaguely remember this in real life when a physical printer would drop a PDF print job at a very small number of pages (10 or so), e.g. when the PDF was full of large, detailed images that consumed a lot of MB's per image.

**Comment by peter-vandenabeele-axa** [comment]
image

**Pull request merged**